src/OrchardCore/Orchard.Mvc.Core/ModuleProjectRazorFileProvider.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Linq;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.FileProviders.Physical;
+using Microsoft.Extensions.Primitives;
+
+namespace Orchard.Mvc
+{
+    /// <summary>
+    /// This custom <see cref="IFileProvider"/> implementation provides the file contents
+    /// of Module Project Razor files while in a development environment.
+    /// </summary>
+    public class ModuleProjectRazorFileProvider : IFileProvider
+    {
+        private const string MappingFilePath = "obj\\ModuleProjectRazorFiles.map";
+
+        private static ConcurrentDictionary<string, string> _paths;
+        private static object _synLock = new object();
+
+        public ModuleProjectRazorFileProvider(string rootPath)
+        {
+            if (_paths == null)
+            {
+                lock (_synLock)
+                {
+                    if (_paths == null)
+                    {
+                        var path = Path.Combine(rootPath, MappingFilePath);
+
+                        if (File.Exists(path))
+                        {
+                            var paths = File.ReadAllLines(path)
+                                .Select(x => x.Split(new[] { '|' }, StringSplitOptions.RemoveEmptyEntries))
+                                .Where(x => x.Count() == 2).ToDictionary(x => x[1].Replace('\\', '/'), x => x[0]);
+
+                            _paths = new ConcurrentDictionary<string, string>(paths);
+                        }
+                        else
+                        {
+                            _paths = new ConcurrentDictionary<string, string>();
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
`Replace()` is to normalize paths, as [done here](https://github.com/aspnet/FileSystem/blob/dev/src/Microsoft.Extensions.FileProviders.Physical/PhysicalFilesWatcher.cs#L302). Razor uses normalized paths, [see here](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Razor/Internal/ViewPath.cs#L37). So, we need to normalize paths which are used as dictionary keys.

Then, for the paths which are used as values, and from which we create new `FileInfo`, on windows both `DirectorySeparatorChar '\'` and `AltDirectorySeparatorChar '/'` can be used. And i assume that e.g on unix paths generated during building will already use the platform specific `/` separator.

Note: on unix `AltDirectorySeparatorChar` is also `/` , since `\` is not a valid separator on Unix